lib/internal/test_runner/test.js

```diff
@@ -0,0 +1,396 @@
+'use strict';
+const {
+  ArrayPrototypePush,
+  ArrayPrototypeShift,
+  Number,
+  ObjectCreate,
+  SafeMap,
+} = primordials;
+const { AsyncResource } = require('async_hooks');
+const {
+  codes: {
+    ERR_TEST_FAILURE,
+  },
+} = require('internal/errors');
+const { TapStream } = require('internal/test_runner/tap_stream');
+const { createDeferredPromise } = require('internal/util');
+const { isUint32 } = require('internal/validators');
+const { clearInterval, setInterval } = require('timers');
+const { bigint: hrtime } = process.hrtime;
+const kCancelledByParent = 'cancelledByParent';
+const kParentAlreadyFinished = 'parentAlreadyFinished';
+const kSubtestsFailed = 'subtestsFailed';
+const kTestCodeFailure = 'testCodeFailure';
+const kDefaultIndent = '    ';
+const kMaxTimeout = 2 ** 31 - 1;
+const noop = () => {};
+
+class TestContext {
+  #test;
+
+  constructor(test) {
+    this.#test = test;
+  }
+
+  diagnostic(message) {
+    this.#test.diagnostic(message);
+  }
+
+  skip(message) {
+    this.#test.skip(message);
+  }
+
+  todo(message) {
+    this.#test.todo(message);
+  }
+
+  test(name, options, fn) {
+    const subtest = this.#test.createSubtest(name, options, fn);
+
+    return subtest.start();
+  }
+}
+
+class Test extends AsyncResource {
+  constructor(options) {
+    super('Test');
+
+    let { fn, name, parent } = options;
+    const { concurrency, skip } = options;
+
+    if (typeof fn !== 'function') {
+      fn = noop;
+    }
+
+    if (typeof name !== 'string' || name === '') {
+      name = fn.name || '<anonymous>';
```

Comment:
`fn.name === 'noop'` if `fn === noop` (so `node -e 'require("test_runner")("")'` produces `ok 1 - noop` instead of `ok 1 - <anonymous>`). One solution would be to set `const noop = FunctionPrototype`.